internal/ebpf/probesfacade/captor/capture.go

```diff
@@ -107,51 +119,110 @@ func Capture(functionSymbol string, cmdArgs []string, opts CaptureOptions) ([]ui
 	// init perf buffer
 	eventsChannel := make(chan []byte)
 	lostChannel := make(chan uint64)
-	rb, err := bpfModule.InitPerfBuf(bpfEventsMap, eventsChannel, lostChannel, 1)
+	pb, err := bpfModule.InitPerfBuf(bpfEventsMap, eventsChannel, lostChannel, 1)
 	if err != nil {
 		return nil, fmt.Errorf("error initializing map (%s) with PerfBuffer: %v", bpfEventsMap, err)
 	}
 
-	// run args that we want to trace
+	return &ebpfSetup{
+		mod:      bpfModule,
+		link:     traceLink,
+		pb:       pb,
+		eventsCh: eventsChannel,
+		lostCh:   lostChannel,
+		opts:     opts,
+		cmd:      cmdArgs,
+	}, nil
+}
+
+// Close closes the ebpf link and module.
+func (ebpf *ebpfSetup) Close() {
+	ebpf.link.Destroy()
+	ebpf.mod.Close()
+}
+
+// Capture collects syscalls from the executed command.
+// Returns values through the given channels.
+func (ebpf *ebpfSetup) Capture(collectedSyscalls chan []uint32, errorCh chan error) {
+	// setting up ticker to dump results
+	// every interval of time.
+	var ticker *time.Ticker
+	var interval time.Duration
+	if ebpf.opts.Interval > 0 {
+		interval = time.Duration(ebpf.opts.Interval) * time.Second
+	} else {
+		// wait what? ~114 years of interval?
+		// yes, so that the ticker will never be triggered.
+		// the value looks reasonably good.
+		interval = time.Duration(1000000) * time.Hour
+	}
+	ticker = time.NewTicker(interval)
+	defer ticker.Stop()
+
 	var wg sync.WaitGroup
 	wg.Add(1)
-	outputCh := make(chan string)
-	errorCh := make(chan string)
-	go executor.Run(cmdArgs, opts.CommandOutput, opts.CommandError, &wg, outputCh, errorCh)
+	cmdStdoutCh := make(chan string)
+	cmdStderrCh := make(chan string)
+	// running command to trace its syscalls
+	go executor.Run(
+		ebpf.cmd,
+		ebpf.opts.CommandOutput,
+		ebpf.opts.CommandError,
+		&wg,
+		cmdStdoutCh,
+		cmdStderrCh,
+	)
 
 	var syscalls []uint32
 	go func() {
 		for {
 			select {
-			case data := <-eventsChannel:
+			case data := <-ebpf.eventsCh:
 				var e event
 				err := binary.Read(bytes.NewBuffer(data), binary.LittleEndian, &e)
 				if err != nil {
-					fmt.Fprintf(os.Stderr, "error reading event: %v\n", err)
+					//fmt.Fprintf(os.Stderr, "error reading event: %v\n", err)
+					// will be left empty for now.
```

Comment:
maybe we can send  this error down errorCh?

Reply:
I could be an idea. I'm wondering if we should stop the execution when a reading fails. This usually happens at the end of the executed command for EOF, but for other cases this could not be appropriate.
I think we can avoid stopping harpoon just for a reading error for now and let it continue the execution.

Reply:
Oh ok I misinterpreted it, I thought it was the cmd error
:+1: